### ISSUE TITLE
Ep 04: Use automatic variable for all prerequisites

### ISSUE THREAD

**Issue by tomichec** [swcarpentry/make-novice#166]
I am currently reviewing the course on the Make. I came across an issue in the dependencies episode.

The recipe
```
results.txt : testzipf.py isles.dat abyss.dat last.dat
	python $< *.dat > $@
```
might work correctly, but it teaches a poor practice. This is because it assumes that no further files `*.dat` are present in the folder, which might not be the case. If there are further files called `*.dat` as it happens often even within small projects, those would be included in the action.

I think a much better approach would be using the automated variable for representing all the pre-requisites, i.e. `$^` or `$+` instead of `$< *.dat`. However, as I understand it `$^` is not guaranteed to contain all the prerequisites listed in the same order (for sure at least the prerequisites, that are defined several times are present only once). The problem with `$+` is that it was not taught in the episode 03 on automatic variables, so many other changes are required in the code of the Makefiles.

**Comment by tomichec** [comment]
Thinking about the whole section, I think that the splitting the concept of dependencies into different categories for data and scripts is really confusing. From the concept of Make there is really no difference if the file is script or data.

I think that already in the explanation of Makefile it should be explained that those files are considered in the same way and if the timestamp of the data or scripts changes, the target will be recreated.

**Comment by gcapes** [comment]
Thanks for raising an issue! I'll try to have a proper look at it this week.

**Comment by tomichec** [comment]
I agree with the first part and can do the changes when I have some time.

For the second one, I see the most advantage of `%` and `$*` in the possibility of passing it as an argument to the function, e.g.
```
all: sin_1.dat sin_2.dat
sin_%.dat:  calculate_sin.py
	python $< --amplitude $* > $@
```
or even
```
all: sin--amplitude_1_--period_1.dat
sin%.dat: calculate_sin.py
	python $< $(subst _, , $*) > $@
```

Otherwise, I see using `$@` more neat.

Now I realise there is a comment about this in the chapter (I haven't advanced that far yet), but perhaps using some command line arguments might be also an option.

**Comment by gcapes** [comment]
I've now split the second point out into a separate issue.

**Comment by gcapes** [comment]
I've started work on these issues, but it looks like it will need a rethink/reorder of the variable/functions episodes in order to accommodate these changes.
